File: NeuronMCP/internal/tools/executor.go
package tools

import (
	"context"
	"encoding/json"
	"fmt"

	"github.com/jackc/pgx/v5"
	"github.com/pgElephant/NeuronMCP/internal/database"
)

// QueryExecutor executes database queries for tools
type QueryExecutor struct {
	db *database.Database
}

// NewQueryExecutor creates a new query executor
func NewQueryExecutor(db *database.Database) *QueryExecutor {
	return &QueryExecutor{db: db}
}

// ExecuteVectorSearch executes a vector search query
func (e *QueryExecutor) ExecuteVectorSearch(ctx context.Context, table, vectorColumn string, queryVector []interface{}, distanceMetric string, limit int, additionalColumns []interface{}) ([]map[string]interface{}, error) {
	// Convert queryVector to []float32
	vec := make([]float32, 0, len(queryVector))
	for _, v := range queryVector {
		if f, ok := v.(float64); ok {
			vec = append(vec, float32(f))
		} else if f, ok := v.(float32); ok {
			vec = append(vec, f)
		} else {
			return nil, fmt.Errorf("invalid vector element type")
		}
	}

	// Convert additional columns to []string
	cols := make([]string, 0, len(additionalColumns))
	for _, col := range additionalColumns {
		if str, ok := col.(string); ok {
			cols = append(cols, str)
		}
	}

	qb := &database.QueryBuilder{}
	query, params := qb.VectorSearch(table, vectorColumn, vec, distanceMetric, limit, cols, nil)

	rows, err := e.db.Query(ctx, query, params...)
	if err != nil {
		return nil, fmt.Errorf("failed to execute vector search: %w", err)
	}
	defer rows.Close()

	return scanRowsToMaps(rows)
}

// ExecuteQuery executes a query and returns all rows
func (e *QueryExecutor) ExecuteQuery(ctx context.Context, query string, params []interface{}) ([]map[string]interface{}, error) {
	rows, err := e.db.Query(ctx, query, params...)
	if err != nil {
		return nil, fmt.Errorf("failed to execute query: %w", err)
	}
	defer rows.Close()

	return scanRowsToMaps(rows)
}

// ExecuteQueryOne executes a query and returns a single row
func (e *QueryExecutor) ExecuteQueryOne(ctx context.Context, query string, params []interface{}) (map[string]interface{}, error) {
	rows, err := e.db.Query(ctx, query, params...)
	if err != nil {
		return nil, fmt.Errorf("failed to execute query: %w", err)
	}
	defer rows.Close()

	if !rows.Next() {
		return nil, fmt.Errorf("no rows returned")
	}

	result, err := scanRowToMap(rows)
	if err != nil {
		return nil, err
	}

	if rows.Next() {
		return nil, fmt.Errorf("multiple rows returned")
	}

	return result, nil
}

// scanRowsToMaps scans all rows into maps
func scanRowsToMaps(rows pgx.Rows) ([]map[string]interface{}, error) {
	var results []map[string]interface{}

	for rows.Next() {
		row, err := scanRowToMap(rows)
		if err != nil {
			return nil, err
		}
		results = append(results, row)
	}

	return results, rows.Err()
}

// scanRowToMap scans a single row into a map
func scanRowToMap(rows pgx.Rows) (map[string]interface{}, error) {
	fieldDescriptions := rows.FieldDescriptions()
	values := make([]interface{}, len(fieldDescriptions))
	valuePointers := make([]interface{}, len(fieldDescriptions))

	for i := range values {
		valuePointers[i] = &values[i]
	}

	if err := rows.Scan(valuePointers...); err != nil {
		return nil, err
	}

	result := make(map[string]interface{})
	for i, desc := range fieldDescriptions {
		val := values[i]
		// Handle byte arrays (JSON, text, etc.)
		if bytes, ok := val.([]byte); ok {
			// Try to parse as JSON
			var jsonVal interface{}
			if err := json.Unmarshal(bytes, &jsonVal); err == nil {
				val = jsonVal
			} else {
				val = string(bytes)
			}
		}
		result[string(desc.Name)] = val
	}

	return result, nil
}

File: NeuronMCP/src/tools/base/executor.ts
/**
 * Query execution helpers for tools
 */

import { Database } from "../../database/connection.js";
import { QueryResult } from "pg";

export class QueryExecutor {
	constructor(private db: Database) {}

	/**
	 * Execute a query and return rows
	 */
	async executeQuery(query: string, params?: any[]): Promise<any[]> {
		const result = await this.db.query(query, params);
		return result.rows;
	}

	/**
	 * Execute a query and return single row
	 */
	async executeQueryOne(query: string, params?: any[]): Promise<any | null> {
		const result = await this.db.query(query, params);
		return result.rows[0] || null;
	}

	/**
	 * Execute a query and return full result
	 */
	async executeQueryFull(query: string, params?: any[]): Promise<QueryResult> {
		return this.db.query(query, params);
	}

	/**
	 * Execute a function call
	 */
	async executeFunction(functionName: string, params: any[]): Promise<any> {
		const paramPlaceholders = params.map((_, i) => `$${i + 1}`).join(", ");
		const query = `SELECT ${this.db.escapeIdentifier(functionName)}(${paramPlaceholders}) AS result`;
		const result = await this.executeQueryOne(query, params);
		return result?.result;
	}

	/**
	 * Execute a function call that returns a table
	 */
	async executeFunctionTable(functionName: string, params: any[]): Promise<any[]> {
		const paramPlaceholders = params.map((_, i) => `$${i + 1}`).join(", ");
		const query = `SELECT * FROM ${this.db.escapeIdentifier(functionName)}(${paramPlaceholders})`;
		return this.executeQuery(query, params);
	}

	/**
	 * Execute vector search query
	 */
	async executeVectorSearch(
		table: string,
		vectorColumn: string,
		queryVector: number[],
		distanceMetric: "l2" | "cosine" | "inner_product" | "l1" | "hamming" | "chebyshev" | "minkowski" = "l2",
		limit: number = 10,
		additionalColumns: string[] = []
	): Promise<any[]> {
		const { query, params } = this.db.queryBuilder.vectorSearch(
			table,
			vectorColumn,
			queryVector,
			distanceMetric,
			limit,
			additionalColumns
		);
		return this.executeQuery(query, params);
	}

	/**
	 * Execute batch operation
	 */
	async executeBatch(queries: Array<{ query: string; params?: any[] }>): Promise<any[][]> {
		const results: any[][] = [];
		for (const { query, params } of queries) {
			const result = await this.executeQuery(query, params);
			results.push(result);
		}
		return results;
	}

	/**
	 * Execute in transaction
	 */
	async executeTransaction<T>(callback: (executor: QueryExecutor) => Promise<T>): Promise<T> {
		return this.db.transaction(async (client) => {
			// Create a temporary executor with the transaction client
			const tempDb = { ...this.db, pool: { query: (text: string, params?: any[]) => client.query(text, params) } } as any;
			const tempExecutor = new QueryExecutor(tempDb);
			return callback(tempExecutor);
		});
	}
}





